### PR TITLE
Fix datetimeDiff for redshift if the datetime arguments are the same

### PR DESCRIPTION
**Pull request by calherries**
This fixes a bug where the `datetimeDiff` custom expression function for redshift will return null/nil for "month" and "quarter" units if the datetime arguments are the same datetime value.

e.g. the expression `datetimeDiff("2021-10-03", "2021-10-03", "month")` would return null, instead of 0.

This is caused by a strange bug in redshift: 
`select extract('day' from a) > extract('day' from b) from <table> limit 1`  returns null if `a = b`, when it should return false. The solution is to wrap this expression in a `coalesce` function ([code](https://github.com/metabase/metabase/pull/27220/files#diff-46e2afe72e4fdc7fda5089fd547af5b814cdf0a451a51f16a8da93dad23aab63R231))

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/27220)

### CODE REVIEW

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj [hunk 1]
```diff
@@ -711,32 +711,40 @@
                                  [:expression "diff-rev"]]})
                      (mt/formatted-rows [int int])
                      first))]
-        (doseq [[unit cases] [[:year [["2021-10-03T09:18:09" "2022-10-02T09:18:09" 0 "day under a year"]
+        (doseq [[unit cases] [[:year [["2021-10-03T09:00:00" "2021-10-03T09:00:00" 0 "same time"]
```
Comment (calherries) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:714]: I added a "same time" test for each unit to catch similar bugs across drivers

**Review by camsaul — APPROVED** [review]
works for me. The change makes sense. Is it the case that `integer > integer` in general returns `NULL` in Redshift if those numbers are equal? I'm trying this out locally and I don't see how that would cause problems elsewhere since we don't really let you do something like `SELECT x > y`... but it might be worth a quick scan thru the code to see if anything else is using `:>` or similar and needs to be fixed.

**Comment by calherries** [comment]
@camsaul 
>  Is it the case that integer > integer in general returns NULL

Thankfully, no.

```
(ffirst (:rows (dev/query-jdbc-db :redshift "select extract(day from cast('2022-01-01' as timestamp)) > extract(day from cast('2022-01-01' as timestamp)) FROM \"schema_164\".\"test_data_venues\"")))
;; => nil
(ffirst (:rows (dev/query-jdbc-db :redshift "select 1 > 1 FROM \"schema_164\".\"test_data_venues\"")))
;; => false
(ffirst (:rows (dev/query-jdbc-db :redshift "select 1 > extract(day from cast('2022-01-01' as timestamp)) FROM \"schema_164\".\"test_data_venues\"")))
;; => false
```

**Comment by calherries** [comment]
Bizarrely, this NULL issue also applies to the `<>` operator but not `=`. I wonder if this could cause an issue with datetime filters in MBQL

(ffirst (:rows (dev/query-jdbc-db :redshift "select extract(day from cast('2022-01-01' as timestamp)) = extract(day from cast('2022-01-01' as timestamp)) FROM \"schema_164\".\"test_data_venues\"")))
;; => true
(ffirst (:rows (dev/query-jdbc-db :redshift "select extract(day from cast('2022-01-01' as timestamp)) <> extract(day from cast('2022-01-01' as timestamp)) FROM \"schema_164\".\"test_data_venues\"")))
;; => nil

**Pull request merged**